distribution.md

```diff
@@ -0,0 +1,180 @@
+\begin{bmatrix}
+O, &amp; P
+\end{bmatrix} = 
+\begin{bmatrix}
+A, &amp; B, &amp; E, &amp; F \\
+A, &amp; B, &amp; G, &amp; H \\
+C, &amp; D, &amp; E, &amp; F \\
+C, &amp; D, &amp; G, &amp; H \\
+\end{bmatrix}
+" />
+
+Note that buildpack IDs are expanded depth-first in left-to-right order.
+
+If a buildpack order entry within a group has the parameter `optional = true`, then a copy of the group without the entry MUST be repeated after the original group.
+
+## Artifact Format
+
+### Buildpack Blob
+
+A buildpack blob MUST contain one or more buildpacks.
+
+A buildpack blob MUST be packaged as gzip-compressed tarball.
+Its filename should end in `.tgz`.
+
+A buildpack blob MUST contain a `buildpack.toml` file at its root directory.
+
+A buildpack defined within `buildpack.toml` MUST either be:
+
+1. A buildpack implementation specified by a `stacks` field or
+2. A buildpack order specified by an `order` field referencing other buildpack IDs/versions.
+
+A buildpack reference in `order` MUST either be:
+
+1. A reference to a buildpack inside of the blob using a `path` field ,
+2. A reference to a buildpack outside of the blob using an `id` and `version` field. 
+
+### Buildpackage
+
+A buildpackage MUST exist as either an OCI image on an image registry, an OCI image in a Docker daemon, or a `.cnb` file.
+
+A `.cnb` file MUST be an uncompressed tar archive containing an OCI image. Its file name SHOULD end in `.cnb`.
+
+Each FS layer blob in the buildpackage MUST contain a single buildpack blob and at least one symlink.
+
+A symlink MUST be created for each buildpack version that the blob is assembled to support.
+
+```
+/cnb/blobs/<sha256 checksum of buildpack blob tgz>/
+/cnb/by-id/<buildpack ID1>/<buildpack version> -> /cnb/blobs/<sha256 checksum of blob tgz>/
+/cnb/by-id/<buildpack ID2>/<buildpack version> -> /cnb/blobs/<sha256 checksum of blob tgz>/<subdir>/
+...
+```
+
+A buildpack ID, buildpack version, and at least one stack MUST be provided in the OCI image config as a Label.
+
+Label: `io.buildpacks.buildpackage.metadata`
+```json
+{
+  "id": "<entrypoint buildpack ID>",
+  "version": "<entrypoint buildpack version>",
+  "stacks": [
+    {
+      "id": "<stack ID>",
+      "mixins": ["<mixin name>"]
+    }
+  ]
+}
+```
+
+The buildpack ID and version MUST match a buildpack provided by a layer blob.
+
+For a buildpackage to be valid, each `buildpack.toml` describing a buildpack implementation MUST have all listed stacks.
+
+For each listed stack, all associated buildpacks MUST be a candidate for detection when the entrypoint buildpack ID and version are selected.
+
+Each stack ID MUST only be present once.
+For a given stack, the `mixins` list MUST enumerate mixins such that no included buildpacks are missing a mixin for the stack.
+
+Fewer stack entries as well as additional mixins for a stack entry MAY be specified.
+
+## Data Format
+
+### buildpack.toml (TOML)
+
+```toml
+[buildpack]
```

Comment:
In order to better determine compatibly between schema changes we should have a `schema` version associated with every spec'd toml file. It would certainly help in the current migration for `buildpack.toml` but would be equally helpful in the future for other `toml` files.

We propose this version be `2` for `builder.toml` so we could assume version `1` for the previous version without this new property.

proposal:
```
schema = "2"
```

CC: @ameyer-pivotal  @ekcasey

Reply:
More examples with some other options:

```toml
schema = "2"

[buildpack]
id = "<buildpack ID>"
name = "<buildpack name>"
version = "<buildpack version>"
```
or 

```toml
version = "2"

[buildpack]
id = "<buildpack ID>"
name = "<buildpack name>"
version = "<buildpack version>"
```

or


```toml
schema-version = "2"

[buildpack]
id = "<buildpack ID>"
name = "<buildpack name>"
version = "<buildpack version>"
```

Reply:
I really like the idea of a schema version. I strongly vote for `schema-version` as the most clear key name.

Reply:
Let's move this to a different PR.

Reply:
@jromero would love to see an RFC for this!

Reply:
@jkutner worth noting that this additional RFC will delay the next lifecycle release significantly. We will likely need to fork the lifecycle to include these changes for the CF builder, and run ahead of the spec.

Reply:
@jkutner RFC here: https://github.com/buildpack/rfcs/pull/19